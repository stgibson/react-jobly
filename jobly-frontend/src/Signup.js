import React, { useState } from "react";
import { useHistory } from "react-router-dom";
import { v4 as uuid } from "uuid";
import Container from "react-bootstrap/Container";
import Form from "react-bootstrap/Form";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import Button from "react-bootstrap/Button";
import Alert from "react-bootstrap/Alert";

/**
 * Component for displaying form for user to create an account
 * @param {Object{function}} param0 
 * @returns JSX code for rendering sign up form
 */
const Signup = ({ signup }) => {
  const history = useHistory();
  const initFormData = {
    username: "",
    password: "",
    firstName: "",
    lastName: "",
    email: ""
  };
  const [formData, setFormData] = useState(initFormData);
  const [errors, setErrors] = useState([]);
  
  /**
   * Updates formData when user types in form
   * @param {Object{any}} evt 
   */
  const handleChange = evt => {
    const { id, value } = evt.target;
    setFormData(formData => ({ ...formData, [id]: value }));
  };

  /**
   * Logs in user and redirects to companies page
   * @param {Object{any}} evt 
   */
  const handleSubmit = async evt => {
    evt.preventDefault();
    const { errors } = await signup(formData)
    if (errors.length) {
      setErrors(errors);
    }
    else {
      history.push("/companies");
    }
  };

  return (
    <Container>
      <div>
        <Row>
          <Col xs={ { span: 6, offset: 3 } }>
            <h3>Sign Up</h3>
            <Form onSubmit={ handleSubmit }>
              <Form.Group controlId="username">
                <Form.Label>Username</Form.Label>
                <Form.Control
                  type="text"
                  onChange={ handleChange }
                  value={ formData.username }
                />
              </Form.Group>
              <Form.Group controlId="password">
                <Form.Label>Password</Form.Label>
                <Form.Control
                  type="password"
                  onChange={ handleChange }
                  value={ formData.password }
                />
              </Form.Group>
              <Form.Group controlId="firstName">
                <Form.Label>First name</Form.Label>
                <Form.Control
                  type="text"
                  onChange={ handleChange }
                  value={ formData.firstName }
                />
              </Form.Group>
              <Form.Group controlId="lastName">
                <Form.Label>Last name</Form.Label>
                <Form.Control
                  type="text"
                  onChange={ handleChange }
                  value={ formData.lastName }
                />
              </Form.Group>
              <Form.Group controlId="email">
                <Form.Label>Email</Form.Label>
                <Form.Control
                  type="text"
                  onChange={ handleChange }
                  value={ formData.email }
                />
              </Form.Group>
              {
                errors.map(error => (
                  <Alert key={ uuid() } variant="danger">{ error }</Alert>
                ))
              }
              <Button variant="primary" type="submit">Submit</Button>
            </Form>
          </Col>
        </Row>
      </div>
    </Container>
  );
};

export default Signup;