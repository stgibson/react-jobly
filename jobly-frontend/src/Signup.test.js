import React from "react";
import { MemoryRouter } from "react-router-dom";
import { render, fireEvent } from "@testing-library/react";
import Signup from "./Signup";

const signup = jest.fn(() => ({ errors: [] })); // Learned how to mock functions at https://jestjs.io/docs/mock-functions

it("renders without crashing", () => {
  render(<MemoryRouter><Signup /></MemoryRouter>);
});

it("matches snapshot", () => {
  const { asFragment } = render(<MemoryRouter><Signup /></MemoryRouter>);
  
  expect(asFragment()).toMatchSnapshot();
});

it("updates username input when user types", () => {
  const { queryByLabelText, queryByDisplayValue } = render(
    <MemoryRouter><Signup /></MemoryRouter>
  );

  const input = queryByLabelText("Username");
  
  fireEvent.change(input, { target: { id: "username", value: "test" } });

  expect(queryByDisplayValue("test")).toBeInTheDocument();
});

it("updates password input when user types", () => {
  const { queryByLabelText, queryByDisplayValue } = render(
    <MemoryRouter><Signup /></MemoryRouter>
  );

  const input = queryByLabelText("Password");
  
  fireEvent.change(input, { target: { id: "password", value: "password" } });

  expect(queryByDisplayValue("password")).toBeInTheDocument();
});

it("updates first name input when user types", () => {
  const { queryByLabelText, queryByDisplayValue } = render(
    <MemoryRouter><Signup /></MemoryRouter>
  );

  const input = queryByLabelText("First name");
  
  fireEvent.change(input, { target: { id: "firstName", value: "Test" } });

  expect(queryByDisplayValue("Test")).toBeInTheDocument();
});

it("updates last name input when user types", () => {
  const { queryByLabelText, queryByDisplayValue } = render(
    <MemoryRouter><Signup /></MemoryRouter>
  );

  const input = queryByLabelText("Last name");
  
  fireEvent.change(input, { target: { id: "lastName", value: "User" } });

  expect(queryByDisplayValue("User")).toBeInTheDocument();
});

it("updates email input when user types", () => {
  const { queryByLabelText, queryByDisplayValue } = render(
    <MemoryRouter><Signup /></MemoryRouter>
  );

  const input = queryByLabelText("Email");
  
  fireEvent.change(
    input,
    { target: { id: "email", value: "test@gmail.com" } }
  );

  expect(queryByDisplayValue("test@gmail.com")).toBeInTheDocument();
});

it("calls signup when user submits form", () => {
  const { queryByLabelText, queryByText } = render(
    <MemoryRouter><Signup signup={ signup } /></MemoryRouter>
  );

  const usernameInput = queryByLabelText("Username");
  const passwordInput = queryByLabelText("Password");
  const firstNameInput = queryByLabelText("First name");
  const lastNameInput = queryByLabelText("Last name");
  const emailInput = queryByLabelText("Email");
  const submitButton = queryByText("Submit");
  
  fireEvent.change(usernameInput, { target: { value: "test" } });
  fireEvent.change(passwordInput, { target: { value: "password" } });
  fireEvent.change(firstNameInput, { target: { id: "firstName", value: "Test" } });
  fireEvent.change(lastNameInput, { target: { id: "lastName", value: "User" } });
  fireEvent.change(
    emailInput,
    { target: { id: "email", value: "test@gmail.com" } }
  );
  fireEvent.click(submitButton);

  expect(signup.mock.calls.length).toEqual(1);
  expect(signup.mock.calls[0][0].username).toEqual("test");
  expect(signup.mock.calls[0][0].password).toEqual("password");
  expect(signup.mock.calls[0][0].firstName).toEqual("Test");
  expect(signup.mock.calls[0][0].lastName).toEqual("User");
  expect(signup.mock.calls[0][0].email).toEqual("test@gmail.com");
});
