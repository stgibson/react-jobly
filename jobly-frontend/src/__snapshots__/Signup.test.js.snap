// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div>
      <div
        class="row"
      >
        <div
          class="col-6 offset-3"
        >
          <h3>
            Sign Up
          </h3>
          <form
            class=""
          >
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="username"
              >
                Username
              </label>
              <input
                class="form-control"
                id="username"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="password"
              >
                Password
              </label>
              <input
                class="form-control"
                id="password"
                type="password"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="firstName"
              >
                First name
              </label>
              <input
                class="form-control"
                id="firstName"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="lastName"
              >
                Last name
              </label>
              <input
                class="form-control"
                id="lastName"
                type="text"
                value=""
              />
            </div>
            <div
              class="form-group"
            >
              <label
                class="form-label"
                for="email"
              >
                Email
              </label>
              <input
                class="form-control"
                id="email"
                type="text"
                value=""
              />
            </div>
            <button
              class="btn btn-primary"
              type="submit"
            >
              Submit
            </button>
          </form>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
